src/hotspot/share/opto/loopUnswitch.cpp

```diff
@@ -222,69 +227,172 @@ bool PhaseIdealLoop::has_control_dependencies_from_predicates(LoopNode* head) co
-  assert(proj_true->is_IfTrue(), "must be true projection");
-  Node* entry = head->skip_strip_mined()->in(LoopNode::EntryControl);
-  Predicates predicates(entry);
-  if (!predicates.has_any()) {
-    // No Parse Predicate.
-    Node* uniqc = proj_true->unique_ctrl_out();
-    assert((uniqc == head && !head->is_strip_mined()) || (uniqc == head->in(LoopNode::EntryControl)
-                                                          && head->is_strip_mined()), "must hold by construction if no predicates");
-  } else {
-    // There is at least one Parse Predicate. When skipping all predicates/predicate blocks, we should end up
-    // at 'proj_true'.
-    assert(proj_true == predicates.entry(), "must hold by construction if at least one Parse Predicate");
+  void verify_unswitched_loop_versions(LoopNode* true_path_loop_head,
+                                       const UnswitchedLoopSelector& unswitched_loop_selector) const {
+    verify_unswitched_loop_version(true_path_loop_head, unswitched_loop_selector.true_path_loop_proj());
+    verify_unswitched_loop_version(old_to_new(true_path_loop_head)->as_Loop(),
+                                   unswitched_loop_selector.false_path_loop_proj());
+  }
+
+  static void verify_unswitched_loop_version(LoopNode* loop_head, IfProjNode* loop_selector_if_proj) {
+    Node* entry = loop_head->skip_strip_mined()->in(LoopNode::EntryControl);
+    const Predicates predicates(entry);
+    // When skipping all predicates, we should end up at 'loop_selector_if_proj'.
+    assert(loop_selector_if_proj == predicates.entry(), "should end up at loop selector If");
   }
-}
 #endif // ASSERT
 
+ public:
+  // Unswitch on the invariant loop selector by creating two unswitched loop versions.
+  void unswitch(const UnswitchedLoopSelector& unswitched_loop_selector) {
+    _phase->clone_loop(_loop, *_old_new, _phase->dom_depth(_strip_mined_loop_head),
+                       PhaseIdealLoop::CloneIncludesStripMined, unswitched_loop_selector.selector());
+
+    // At this point, the selector If projections are the corresponding loop entries.
+    // clone_parse_and_assertion_predicates_to_unswitched_loop() could clone additional predicates after the selector
+    // If projections. The loop entries are updated accordingly.
+    IfProjNode* true_path_loop_entry = unswitched_loop_selector.true_path_loop_proj();
+    IfProjNode* false_path_loop_entry = unswitched_loop_selector.false_path_loop_proj();
+    _phase->clone_parse_and_assertion_predicates_to_unswitched_loop(_loop, *_old_new,
+                                                                    true_path_loop_entry, false_path_loop_entry);
+
+    fix_loop_entries(true_path_loop_entry, false_path_loop_entry);
+
+    DEBUG_ONLY(verify_unswitched_loop_versions(_loop->_head->as_Loop(), unswitched_loop_selector);)
+  }
+};
+
+// Create a true-path- and a false-path-loop version of the original loop by cloning the loop and inserting an If to
```

Comment:
Don't you already have the selector if at this point? The comment suggests you are inserting it here.